File: py/d2p1.py
#!/usr/bin/python3

from fileinput import input


WIN = {
    'A': { 'X': 3, 'Y': 6, 'Z': 0 },
    'B': { 'X': 0, 'Y': 3, 'Z': 6 },
    'C': { 'X': 6, 'Y': 0, 'Z': 3 },
}
SHAPE = { 'X': 1, 'Y': 2, 'Z': 3 }

score = 0
for op, _, me, _ in input():
    score += WIN[op][me] + SHAPE[me]
print(score)

File: src/bin/d2p1.rs
use std::io::{self, BufRead};
use RPS::*;

#[derive(Clone, Copy)]
enum RPS {
    Rock,
    Paper,
    Scisors,
}

impl RPS {
    fn from_byte(byte: u8) -> RPS {
        match byte {
            b'A' | b'X' => Rock,
            b'B' | b'Y' => Paper,
            b'C' | b'Z' => Scisors,
            _ => unreachable!("invalid input"),
        }
    }
}

fn main() {
    let mut stdin = io::stdin().lock();
    let mut buf = Vec::new();
    let mut score = 0u64;

    while stdin.read_until(b'\n', &mut buf).unwrap() > 0 {
        let &[op, _, me, _] = buf.as_slice() else { panic!("invalid input"); };
        let (op, me) = (RPS::from_byte(op), RPS::from_byte(me));
        buf.clear();

        let win = match (me, op) {
            (Rock, Scisors) | (Paper, Rock) | (Scisors, Paper) => 6, // win
            (Rock, Rock) | (Paper, Paper) | (Scisors, Scisors) => 3, // draw
            (Rock, Paper) | (Paper, Scisors) | (Scisors, Rock) => 0, // loss
        };

        let shape = match me {
            Rock => 1,
            Paper => 2,
            Scisors => 3,
        };

        score += win + shape;
    }

    println!("{score}");
}
